storage/series/series.go

```diff
@@ -433,52 +440,61 @@ func (s *dbSeries) addBlockWithLock(b block.DatabaseBlock) {
+func (s *dbSeries) Bootstrap(bootstrappedBlocks block.DatabaseSeriesBlocks) (BootstrapResult, error) {
 	s.Lock()
-	defer s.Unlock()
+	defer s.bootstrapComplete()
 
+	var result BootstrapResult
 	if s.bs == bootstrapped {
-		return nil
+		return result, errSeriesAlreadyBootstrapped
 	}
-	if s.bs == bootstrapping {
-		return errSeriesIsBootstrapping
+
+	if bootstrappedBlocks == nil {
+		return result, nil
 	}
 
-	s.bs = bootstrapping
-	existingBlocks := s.blocks
-
-	multiErr := xerrors.NewMultiError()
-	if blocks == nil {
-		// If no data to bootstrap from then fallback to the empty blocks map
-		blocks = existingBlocks
-	} else {
-		// Request the in-memory buffer to drain and reset so that the start times
-		// of the blocks in the buckets are set to the latest valid times
-		s.buffer.DrainAndReset()
-
-		// If any received data falls within the buffer then we emplace it there
-		min, _ := s.buffer.MinMax()
-		for tNano, block := range blocks.AllBlocks() {
-			t := tNano.ToTime()
-			if !t.Before(min) {
-				if err := s.buffer.Bootstrap(block); err != nil {
-					multiErr = multiErr.Add(s.newBootstrapBlockError(block, err))
-				}
-				blocks.RemoveBlockAt(t)
+	// Request the in-memory buffer to drain and reset so that the start times
+	// of the blocks in the buckets are set to the latest valid times
+	s.buffer.DrainAndReset()
+	min, _, err := s.buffer.MinMax()
+	if err != nil {
+		return result, err
+	}
+
+	var (
+		multiErr = xerrors.NewMultiError()
+	)
+	for tNano, block := range bootstrappedBlocks.AllBlocks() {
+		t := tNano.ToTime()
+		// If there is a writable, undrained series buffer bucket then store the block
+		// there and it will be merged / drained as part of the usual lifecycle.
+		if !t.Before(min) {
+			if err := s.buffer.Bootstrap(block); err != nil {
+				multiErr = multiErr.Add(s.newBootstrapBlockError(block, err))
 			}
+			result.NumBlocksMovedToBuffer++
+			continue
 		}
 
-		// If we're overwriting the blocks then merge any existing blocks
-		// already drained
-		for _, existingBlock := range existingBlocks.AllBlocks() {
-			s.mergeBlockWithLock(blocks, existingBlock)
+		// If we're unable to put the blocks in an active series buffer bucket, then store them
+		// in the series block, merging with any existing blocks if necessary. There could be an
+		// existing block in the situation that currentTime > blockStart.Add(blockSize).Add(bufferPast),
+		// in which case the series buffer buckets may have been drained and rotated into a block, but
+		// still exist in memory because a flush hasn't occurred yet.
+		err := s.mergeBlockWithLock(block)
+		if err != nil {
+			multiErr = multiErr.Add(s.newBootstrapBlockError(block, err))
 		}
+		result.NumBlocksMerged++
 	}
 
-	s.blocks = blocks
 	s.bs = bootstrapped
+	return result, multiErr.FinalError()
+}
 
-	return multiErr.FinalError()
+func (s *dbSeries) bootstrapComplete() {
+	s.Unlock()
+	s.bs = bootstrapped
```

Comment:
doesn't the order of these operations need to be reversed?

Reply:
unrelated to your change, but while you're here, mind changing `bs` to `bootstrapState` or some such

Reply:
Yeah that was my bad, I wrote some inline code as a suggestion and wasn't thinking was just spewing it out.  Unlock needs to come last indeed.